lld/test/wasm/lto/stub-library.s

```diff
@@ -1,6 +1,10 @@
 ## The function `bar` is declared in stub.so and depends on `foo` which is
 ## defined in an LTO object.  We also test the case where the LTO object is
 ## with an archive file.
+## The function `baz` is declared in stub.so and depends on `quux`, and both
+## `baz` and `quux` are defined in an LTO object. When `baz` is DCE'd and
+## becomes undefined in the LTO process, `quux` should still be preserved and
+## exported.
```

Comment:
I'm not a little confused I'm afarid.

If baz is never actually undefined (i.e. never actually needs to be imported at all) why would we want to use the import from `stub.so` and therefore why would we ever export quux in this case?  

if baz was used in this program wouldn't we pick the one defined internally rather than externally?  But baz is never actually needed anyway.

Reply:
I'm actually not sure myself why we should preserve and export it, but it looks that's what `processStubLibraries` expects:
https://github.com/llvm/llvm-project/blob/2040b55b772a780f4e7023a79fe0551d96c19e5a/lld/wasm/Driver.cpp#L1108-L1109

`baz` is DCE'd and becomes undefined, but `processStubLibraries` tries to export the deps in case `sym && sym->isUndefined()`. My fix is a way of avoiding this crash. Should we fix this part instead?

Reply:
It was defined and pulled in to LTO but was removed and became undefined in https://github.com/llvm/llvm-project/blob/24297bea9672722d8fbaaff137b301b0becaae9c/llvm/lib/LTO/LTO.cpp#L1042-L1054 because it was dead. And after the LTO `processStubLibraries` sees it  as undefined and decides it should be imported

> The stub library symbol should not be used in the case that a definition exists in the program being linked.

I agree but that's not how the code here is written. We should either (unnecessarily) make the deps available and export them (this PR) or exclude these cases in `processStubLibraries` after LTO. I think the latter is better than this PR (because it avoids unnecessary exports) Maybe I should try that (even though I'm not sure how to do it now; I can investigate)